File: web/src/filter.ts
// TypeScript port of the Live Tail filter grammar (spec.md ss27,
// internal/query/filter.go). Whitespace-separated terms are ANDed. A term
// is either free text (substring match on service_name / root_span_name)
// or key:value / key:op value (op in > >= < <=, meaningful for duration).
//
// This mirrors the backend's behavior for the keys the browser can
// evaluate against a TraceSummary (status, duration, service.name/service,
// free text). TraceSummary carries only the root span's name, not every
// span name, and no attributes, so free-text and arbitrary attribute keys
// are necessarily an approximation of the server-side match — a soft
// no-op for keys we can't evaluate client-side, matching the grammar's
// "never fail, never crash" contract.
import type { TraceSummary } from "./types";

interface Term {
  key: string;
  op: "" | ":" | ">" | ">=" | "<" | "<=";
  text: string;
}

export type Filter = Term[];

const DURATION_OPS: Array<Term["op"]> = [">=", "<=", ">", "<"];

export function parseFilter(query: string): Filter {
  return query
    .split(/\s+/)
    .filter((s) => s.length > 0)
    .map(parseTerm);
}

function parseTerm(raw: string): Term {
  const idx = raw.indexOf(":");
  if (idx < 0) return { key: "", op: "", text: raw.toLowerCase() };

  const key = raw.slice(0, idx).toLowerCase();
  let val = raw.slice(idx + 1);
  let op: Term["op"] = ":";
  for (const candidate of DURATION_OPS) {
    if (val.startsWith(candidate)) {
      op = candidate;
      val = val.slice(candidate.length);
      break;
    }
  }
  return { key, op, text: val };
}

export function matchesFilter(t: TraceSummary, filter: Filter): boolean {
  return filter.every((term) => matchTerm(t, term));
}

function matchTerm(t: TraceSummary, term: Term): boolean {
  if (term.key === "") return matchFreeText(t, term.text);

  switch (term.key) {
    case "status":
      return t.has_error === (term.text.toLowerCase() === "error");
    case "duration":
      return matchDuration(t, term.op, term.text);
    case "service.name":
    case "service":
      return t.service_name.toLowerCase().includes(term.text.toLowerCase());
    default:
      // Unrecognized / attribute keys: TraceSummary carries no span
      // attributes, so treat as a no-op (spec.md ss27).
      return true;
  }
}

function matchFreeText(t: TraceSummary, text: string): boolean {
  return (
    t.service_name.toLowerCase().includes(text) ||
    t.root_span_name.toLowerCase().includes(text)
  );
}

// parseGoDuration parses a Go-style duration string ("1s", "500ms", "1.5m")
// into milliseconds, matching time.ParseDuration's units well enough for
// the durations traces realistically span.
function parseGoDuration(text: string): number | null {
  const match = /^([0-9]*\.?[0-9]+)(ns|us|µs|ms|s|m|h)$/.exec(text.trim());
  if (!match) return null;
  const value = parseFloat(match[1]);
  const unit = match[2];
  const perMs: Record<string, number> = {
    ns: 1e-6,
    us: 1e-3,
    "µs": 1e-3,
    ms: 1,
    s: 1000,
    m: 60_000,
    h: 3_600_000,
  };
  return value * perMs[unit];
}

function matchDuration(t: TraceSummary, op: Term["op"], text: string): boolean {
  const want = parseGoDuration(text);
  if (want === null) return false;
  const got = t.duration_ms;
  switch (op) {
    case ">":
      return got > want;
    case ">=":
      return got >= want;
    case "<":
      return got < want;
    case "<=":
      return got <= want;
    default:
      return got === want;
  }
}

File: internal/query/filter.go
// Package query implements the Live Tail UI filter (spec.md ss27). This is
// purely a display filter: it never influences the Sampling Decision.
package query

import (
	"strconv"
	"strings"
	"time"

	"github.com/ucpr/tailsamplingpreviewer/internal/trace"
)

// Filter is a parsed Live Tail query, e.g. `service.name:payment
// status:error duration:>1s`. All terms are ANDed together.
type Filter struct {
	terms []term
}

type term struct {
	key  string // "" for a free-text term
	op   string // "", ":", ">", ">=", "<", "<="
	text string
}

// Parse builds a Filter from raw query text. Parse never fails: malformed
// terms are treated as free-text substring matches, matching the forgiving
// feel of a Datadog-style search bar (spec.md ss27).
func Parse(query string) Filter {
	fields := strings.Fields(query)
	terms := make([]term, 0, len(fields))
	for _, f := range fields {
		terms = append(terms, parseTerm(f))
	}
	return Filter{terms: terms}
}

func parseTerm(f string) term {
	idx := strings.IndexByte(f, ':')
	if idx < 0 {
		return term{text: strings.ToLower(f)}
	}
	key := strings.ToLower(f[:idx])
	val := f[idx+1:]

	op := ":"
	for _, candidate := range []string{">=", "<=", ">", "<"} {
		if strings.HasPrefix(val, candidate) {
			op = candidate
			val = val[len(candidate):]
			break
		}
	}
	return term{key: key, op: op, text: val}
}

// Empty reports whether the filter has no terms (matches everything).
func (f Filter) Empty() bool { return len(f.terms) == 0 }

// Match reports whether t satisfies every term in the filter.
func (f Filter) Match(t *trace.Trace) bool {
	for _, term := range f.terms {
		if !term.match(t) {
			return false
		}
	}
	return true
}

func (t term) match(tr *trace.Trace) bool {
	if t.key == "" {
		return matchFreeText(tr, t.text)
	}

	switch t.key {
	case "status":
		want := strings.EqualFold(t.text, "error")
		return tr.HasError() == want
	case "duration":
		return matchDuration(tr, t.op, t.text)
	case "service.name", "service":
		return strings.Contains(strings.ToLower(tr.RootServiceName()), strings.ToLower(t.text))
	default:
		return matchAttribute(tr, t.key, t.text)
	}
}

func matchFreeText(tr *trace.Trace, text string) bool {
	if strings.Contains(strings.ToLower(tr.RootServiceName()), text) {
		return true
	}
	for _, s := range tr.Spans {
		if strings.Contains(strings.ToLower(s.Name), text) {
			return true
		}
	}
	return false
}

func matchDuration(tr *trace.Trace, op, text string) bool {
	want, err := time.ParseDuration(text)
	if err != nil {
		return false
	}
	got := tr.Duration()
	switch op {
	case ">":
		return got > want
	case ">=":
		return got >= want
	case "<":
		return got < want
	case "<=":
		return got <= want
	default: // ":"
		return got == want
	}
}

func matchAttribute(tr *trace.Trace, key, text string) bool {
	for _, s := range tr.Spans {
		v, ok := s.Attributes[key]
		if !ok {
			continue
		}
		if str, ok := v.(string); ok && strings.EqualFold(str, text) {
			return true
		}
		if str := toDisplayString(v); strings.EqualFold(str, text) {
			return true
		}
	}
	return false
}

func toDisplayString(v any) string {
	switch x := v.(type) {
	case string:
		return x
	case bool:
		return strconv.FormatBool(x)
	case float64:
		return strconv.FormatFloat(x, 'f', -1, 64)
	case int64:
		return strconv.FormatInt(x, 10)
	default:
		return ""
	}
}
